### PR TITLE
django-digest redis cache nonce storage backend

### PR DESCRIPTION
**Pull request by bufke**
This change solves a deadlock issue from the django-digest database nonce storage backend. django-digest can potentially delete a lot of old nonces all at once, while storing a new one. If this happens while an update occurs, which happens when checking an existing nonce, it deadlocks.

This redis based backend fixes this by storing nonces in redis with a TTL. No additional nonce cleanup is necessary. It requires a blocking redis lock to prevent replay attacks. The lock has sane timeouts applied. Because nonces are small and transient, in memory storage is appropriate. Nonces can also be deleted at any time without apparent consequence (I'm still verifying this claim). It would be fine to store these in a ephemeral redis instance if desired.

Introduces new default cache backend that matches KPI implementation. Only django-redis is supported due to the need for redis locking.

**This PR requires a change to kobo-install to set the new CACHE_URL environment variable**

I believe this PR is finished, but I'd like to test it further before marking as ready.

### CODE REVIEW

**Comment by jnm** [comment]
`CACHE_URL` seems to be added to kobo-install by https://github.com/kobotoolbox/kobo-install/pull/193

File: onadata/apps/django_digest_backends/cache.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        if nonce_count == None:  # No need to lock
+            existing = cache.get(cache_key)
+            if existing is None:
+                return False
+            cache.set(cache_key, NONCE_NO_COUNT, NONCE_TIMEOUT)
+        else:
+            with cache.lock(
+                f'user_nonce_lock_{user}_{nonce}',
+                timeout=NONCE_TIMEOUT,
+                blocking_timeout=30
```
Comment (jnm) [inline onadata/apps/django_digest_backends/cache.py:36]: just curious how this works. is this right?
1. process a takes the lock and somehow dies without releasing it
2. process b tries to get the lock and fails with a `LockError` after 30 seconds
3. process c tries to get the lock and fails with a `LockError` after 30 seconds
4. etc., until the TTL of the lock key (set to `NONCE_TIMEOUT`) expires?

if so, could we catch `LockError` and delete the nonce outright to get back to a better state and minimize the number of worker processes stuck waiting for the lock?

**Review by jnm — CHANGES_REQUESTED** [review]

File: onadata/apps/django_digest_backends/cache.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+DIGEST_NONCE_CACHE_NAME = get_setting('DIGEST_NONCE_CACHE_NAME', 'default')
+NONCE_TIMEOUT = get_setting('DIGEST_NONCE_TIMEOUT_IN_SECONDS', 5 * 60)
+NONCE_NO_COUNT = ''  # Needs to be something other than None to determine not set vs set to null
+
+
+class RedisCacheNonceStorage():
+    def _get_cache(self):
+        # Dynamic fetching of cache is necessary to work with override_settings
```
[the remark this answers is not in the document]
Comment (bufke) [reply onadata/apps/django_digest_backends/cache.py:12]: Originally I made use of this to run with the faster in (local) memory cache. As I added django-redis features, I no longer had tests that made use of this. It seems reasonable to ensure override_settings work as expected, so I kept this.

File: onadata/apps/django_digest_backends/cache.py (hunk 1 above)
Comment (bufke) [reply onadata/apps/django_digest_backends/cache.py]: Implemented. Give this a test please. I used KoboCollect and deleted the nonces from redis. From earlier testing, it seems like it doesn't matter to enketo or collect if we delete the nonces.

**Review by jnm — APPROVED** [review]

**Pull request merged**